### ISSUE TITLE
F1 compute metric warning

### ISSUE THREAD

**Issue by sivakhno** [PyTorchLightning/metrics#811]
## 🐛 Bug

Get a warning in F1 metric compute after following documentation: torchmetrics/utilities/prints.py:36: UserWarning: The ``compute`` method of metric F1 was called before the ``update`` method which may lead to errors, as metric states have not yet been updated. I checked docs and tutorials carefully and all looks correct to me - any suggestions would be most appreciated! Many thanks for help.

### To Reproduce

Error happens in the line `f1 = self.valid_f1.compute()` below 

#### Code sample

```py
class LSTMTaggerModel(pl.LightningModule):
    def __init__(
        self,
        num_classes,
        class_map,
        from_checkpoint=False,
        model_name='last.ckpt',
        learning_rate=3e-6,
        **kwargs,
    ):

        super().__init__()
        self.save_hyperparameters()
        self.learning_rate = learning_rate
        self.model = LSTMClassifier(num_classes=num_classes)
        self.class_map = class_map
        self.num_classes = num_classes
        self.valid_acc = torchmetrics.Accuracy()
        self.valid_f1 = torchmetrics.F1()


    def forward(self, *input, **kwargs):
        return self.model(*input, **kwargs)

    def training_step(self, batch, batch_idx):
        x, y_true = batch
        loss, _ = self(x, labels=y_true)
        self.log('train_loss', loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y_true = batch
        _, y_pred = self(x, labels=y_true)
        preds = torch.argmax(y_pred, axis=1)
        self.valid_acc(preds, y_true[0])
        self.log('val_acc', self.valid_acc, prog_bar=True)
        self.valid_f1(preds, y_true[0])
        self.log('f1', self.valid_f1, prog_bar=True)     

    def configure_optimizers(self):
        'Prepare optimizer and schedule (linear warmup and decay)'
        opt = torch.optim.Adam(params=self.parameters(), lr=self.learning_rate)
        sch = torch.optim.lr_scheduler.CosineAnnealingLR(opt, T_max=10)
        return [opt], [sch]

    def training_epoch_end(self, training_step_outputs):
        avg_loss = torch.tensor([x['loss']
                                 for x in training_step_outputs]).mean()
        self.log('train_loss', avg_loss)
        print(f'###score: train_loss### {avg_loss}')

    def validation_epoch_end(self, val_step_outputs):
        acc = self.valid_acc.compute()
        f1 = self.valid_f1.compute()
        self.log('val_score', acc)
        self.log('f1', f1)
        print(f'###score: val_score### {acc}')

```

### Expected behavior

No warning is provided 

### Environment

- PyTorch Version (e.g., 1.0):
torchmetrics = "0.3.2"
pytorch-lightning = "1.2.7"
torch = "1.7.1"
- OS (e.g., Linux): MacOS
- How you installed PyTorch: with poetry 
- Python version: 3.7
- CUDA/cuDNN version: NA (running on cpu)
- GPU models and configuration: Apple Silicon M1

**Comment by justusschock** [comment]
Hi @sivakhno 

I think the issue is that you log it two times (once in your validation_step and once in the _epoch_end). This means, that it will likely be resetted already in the epoch_end. Can you try the following:

```python
class LSTMTaggerModel(pl.LightningModule):
    def __init__(
        self,
        num_classes,
        class_map,
        from_checkpoint=False,
        model_name='last.ckpt',
        learning_rate=3e-6,
        **kwargs,
    ):

        super().__init__()
        self.save_hyperparameters()
        self.learning_rate = learning_rate
        self.model = LSTMClassifier(num_classes=num_classes)
        self.class_map = class_map
        self.num_classes = num_classes
        self.valid_acc = torchmetrics.Accuracy()
        self.valid_f1 = torchmetrics.F1()


    def forward(self, *input, **kwargs):
        return self.model(*input, **kwargs)

    def training_step(self, batch, batch_idx):
        x, y_true = batch
        loss, _ = self(x, labels=y_true)
        self.log('train_loss', loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y_true = batch
        _, y_pred = self(x, labels=y_true)
        preds = torch.argmax(y_pred, axis=1)
        self.valid_acc(preds, y_true[0])
        self.log('val_acc', self.valid_acc, prog_bar=True, on_epoch=True, on_step=True)
        self.valid_f1(preds, y_true[0])
        self.log('f1', self.valid_f1, prog_bar=True, on_epoch=True, on_step=True)     

    def configure_optimizers(self):
        'Prepare optimizer and schedule (linear warmup and decay)'
        opt = torch.optim.Adam(params=self.parameters(), lr=self.learning_rate)
        sch = torch.optim.lr_scheduler.CosineAnnealingLR(opt, T_max=10)
        return [opt], [sch]

    def training_epoch_end(self, training_step_outputs):
        avg_loss = torch.tensor([x['loss']
                                 for x in training_step_outputs]).mean()
        self.log('train_loss', avg_loss)
        print(f'###score: train_loss### {avg_loss}')
```

**Comment by sivakhno** [comment]
@justusschock - many thanks for fast reply and suggestions. I've used your code but still get the same warnings. They persist even if I turn off (False) on_epoch or on_step flags.

**Comment by SkafteNicki** [comment]
@sivakhno did you remove the `validation_epoch_end` method from your own example when you tried the solution of @justusschock. I am pretty sure that both logging the metric object `self.log(self.metric)` in validation step and logging the computed value `self.log(self.metric.compute())` is what gives the warning. When the metric object is logged, lightning will take care of resetting the metric for you. If this is done before `validation_epoch_end` then the metric is correctly reporting that `update` have not been called (because it was just reset) before `compute`.

This would be the way I would implement it (which does not rely on logging the metric object):
```python
class LSTMTaggerModel(pl.LightningModule):
    def __init__(
        self,
        num_classes,
        class_map,
        from_checkpoint=False,
        model_name='last.ckpt',
        learning_rate=3e-6,
        **kwargs,
    ):

        super().__init__()
        self.save_hyperparameters()
        self.learning_rate = learning_rate
        self.model = LSTMClassifier(num_classes=num_classes)
        self.class_map = class_map
        self.num_classes = num_classes
        self.valid_acc = torchmetrics.Accuracy()
        self.valid_f1 = torchmetrics.F1()


    def forward(self, *input, **kwargs):
        return self.model(*input, **kwargs)

    def training_step(self, batch, batch_idx):
        x, y_true = batch
        loss, _ = self(x, labels=y_true)
        self.log('train_loss', loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y_true = batch
        _, y_pred = self(x, labels=y_true)
        preds = torch.argmax(y_pred, axis=1)
        batch_acc = self.valid_acc(preds, y_true[0])
        self.log('batch_val_acc', batch_acc, prog_bar=True)
        batch_f1 = self.valid_f1(preds, y_true[0])
        self.log('batch_val_f1', batch_f1, prog_bar=True)     

    def configure_optimizers(self):
        'Prepare optimizer and schedule (linear warmup and decay)'
        opt = torch.optim.Adam(params=self.parameters(), lr=self.learning_rate)
        sch = torch.optim.lr_scheduler.CosineAnnealingLR(opt, T_max=10)
        return [opt], [sch]

    def training_epoch_end(self, training_step_outputs):
        avg_loss = torch.tensor([x['loss']
                                 for x in training_step_outputs]).mean()
        self.log('train_loss', avg_loss)
        print(f'###score: train_loss### {avg_loss}')

    def validation_epoch_end(self, val_step_outputs):
        acc = self.valid_acc.compute()
        f1 = self.valid_f1.compute()
        self.log('epoch_val_acc', acc)
        self.log('epoch_val_f1', f1)
        print(f'###score: val_score### {acc}')
```
I assume that you both want the score per batch and accumulated over the epoch.

**Comment by Borda** [comment]
@sivakhno would you be interested in sending a PR to docs update upon you learning from this discussion? :rabbit:
